{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "from dcicutils import ff_utils\n",
    "from functions.notebook_functions import *\n",
    "\n",
    "# get key from file\n",
    "my_key = get_key('koray_data')\n",
    "\n",
    "# all usable env names\n",
    "all_envs = ['data', 'staging', 'fourfront-webdev', 'fourfront-mastertest', 'fourfront-hotseat']\n",
    "# get admin key from s3\n",
    "#my_key = ff_utils.get_authentication_with_server({}, ff_env = 'data')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "245\n"
     ]
    }
   ],
   "source": [
    "q = '/search/?award.project=4DN&experimentset_type=replicate&lab.display_title=Tom+Misteli%2C+NCI&type=ExperimentSetReplicate'\n",
    "gen = ff_utils.search_metadata(q, key=my_key)\n",
    "print(len(gen))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "dict_keys(['aliases', 'link_id', '@id', 'number_of_experiments', 'date_created', 'display_title', 'publications_of_set', 'submitted_by', 'description', 'replicate_exps', 'external_references', 'experiments_in_set', 'status', 'award', 'last_modified', 'principals_allowed', 'uuid', 'experimentset_type', 'static_headers', 'schema_version', 'lab', 'accession', '@type'])\n"
     ]
    }
   ],
   "source": [
    "for i in gen:\n",
    "    print(i.keys())\n",
    "    break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "75\n"
     ]
    }
   ],
   "source": [
    "q = '/search/?lab.display_title=Tom+Misteli%2C+NCI&type=Target'\n",
    "gen = ff_utils.search_metadata(q, key=my_key)\n",
    "print(len(gen))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "prob_num = '''11\n",
    "52\n",
    "91\n",
    "249\n",
    "255\n",
    "433\n",
    "441\n",
    "994\n",
    "74\n",
    "80\n",
    "89\n",
    "232\n",
    "254\n",
    "422\n",
    "385\n",
    "354\n",
    "360\n",
    "389\n",
    "376\n",
    "482\n",
    "463\n",
    "476\n",
    "121\n",
    "263\n",
    "195\n",
    "142\n",
    "146\n",
    "256\n",
    "40\n",
    "156\n",
    "278\n",
    "206\n",
    "86\n",
    "123\n",
    "133\n",
    "163\n",
    "303\n",
    "250\n",
    "141\n",
    "296\n",
    "89\n",
    "193\n",
    "11\n",
    "186\n",
    "157\n",
    "167\n",
    "607\n",
    "609\n",
    "610\n",
    "611\n",
    "613\n",
    "614\n",
    "614.5\n",
    "615\n",
    "616\n",
    "617\n",
    "618\n",
    "220\n",
    "221\n",
    "222\n",
    "223\n",
    "224\n",
    "225\n",
    "225.5\n",
    "226\n",
    "226.5\n",
    "227\n",
    "228\n",
    "229\n",
    "230\n",
    "231'''"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1 \t 2370445 \t 2570720 \t misteli:target_11\n",
      "1 \t 12708719 \t 12865743 \t misteli:target_52\n",
      "1 \t 22223362 \t 22394657 \t misteli:target_91\n",
      "1 \t 61560943 \t 61746906 \t misteli:target_249\n",
      "1 \t 63059760 \t 63224942 \t misteli:target_255\n",
      "1 \t 107482705 \t 107645471 \t misteli:target_433\n",
      "1 \t 109469002 \t 109662490 \t misteli:target_441\n",
      "1 \t 248125293 \t 248263553 \t misteli:target_994\n",
      "1 \t 18566409 \t 18766880 \t misteli:target_74\n",
      "1 \t 20012369 \t 20177632 \t misteli:target_80\n",
      "1 \t 22000503 \t 22169684 \t misteli:target_89\n",
      "1 \t 58032482 \t 58191458 \t misteli:target_232\n",
      "1 \t 63533917 \t 63718092 \t misteli:target_254\n",
      "1 \t 105256970 \t 105427280 \t misteli:target_422\n",
      "1 \t 96067992 \t 96226766 \t misteli:target_385\n",
      "1 \t 88222722 \t 88418471 \t misteli:target_354\n",
      "1 \t 89759075 \t 89913486 \t misteli:target_360\n",
      "1 \t 97025755 \t 97207860 \t misteli:target_389\n",
      "1 \t 93528777 \t 93687930 \t misteli:target_376\n",
      "1 \t 148972721 \t 149165332 \t misteli:target_482\n",
      "1 \t 115549997 \t 115729561 \t misteli:target_463\n",
      "1 \t 118815746 \t 118989848 \t misteli:target_476\n",
      "17 \t 30088606 \t 30263568 \t misteli:target_121\n",
      "17 \t 65509811 \t 65678200 \t misteli:target_263\n",
      "17 \t 48502126 \t 48689497 \t misteli:target_195\n",
      "17 \t 35268046 \t 35441605 \t misteli:target_142\n",
      "17 \t 36387556 \t 36574537 \t misteli:target_146\n",
      "17 \t 63767294 \t 63927790 \t misteli:target_256\n",
      "17 \t 9734635 \t 9900986 \t misteli:target_40\n",
      "17 \t 38743489 \t 38939529 \t misteli:target_156\n",
      "17 \t 69261617 \t 69425619 \t misteli:target_278\n",
      "17 \t 51274882 \t 51467771 \t misteli:target_206\n",
      "17 \t 21275223 \t 21478206 \t misteli:target_86\n",
      "17 \t 30541949 \t 30739347 \t misteli:target_123\n",
      "17 \t 33020947 \t 33186327 \t misteli:target_133\n",
      "18 \t 40596626 \t 40761745 \t misteli:target_163\n",
      "18 \t 75525179 \t 75695249 \t misteli:target_303\n",
      "18 \t 62276743 \t 62456847 \t misteli:target_250\n",
      "18 \t 34992711 \t 35151964 \t misteli:target_141\n",
      "18 \t 73778050 \t 73945077 \t misteli:target_296\n",
      "1 \t 22000503 \t 22169684 \t misteli:target_89\n",
      "18 \t 48016594 \t 48189106 \t misteli:target_193\n",
      "1 \t 2370445 \t 2570720 \t misteli:target_11\n",
      "18 \t 46316193 \t 46486760 \t misteli:target_186\n",
      "18 \t 39016717 \t 39206739 \t misteli:target_157\n",
      "18 \t 41537838 \t 41727804 \t misteli:target_167\n",
      "4 \t 150978277 \t 151124968 \t misteli:target_607\n",
      "4 \t 151276234 \t 151444375 \t misteli:target_609\n",
      "4 \t 151542617 \t 151679223 \t misteli:target_610\n",
      "4 \t 151942840 \t 152102227 \t misteli:target_611\n",
      "4 \t 152319856 \t 152499280 \t misteli:target_613\n",
      "4 \t 152521628 \t 152677396 \t misteli:target_614\n",
      "4 \t 152712784 \t 152899106 \t misteli:target_614_5\n",
      "4 \t 152897610 \t 153104002 \t misteli:target_615\n",
      "4 \t 153155365 \t 153328553 \t misteli:target_616\n",
      "4 \t 153347384 \t 153505952 \t misteli:target_617\n",
      "4 \t 153567330 \t 153719454 \t misteli:target_618\n",
      "4 \t 54270488 \t 54435016 \t misteli:target_220\n",
      "4 \t 54542457 \t 54695030 \t misteli:target_221\n",
      "4 \t 54794388 \t 54995522 \t misteli:target_222\n",
      "4 \t 55053142 \t 55196692 \t misteli:target_223\n",
      "4 \t 55236649 \t 55437873 \t misteli:target_224\n",
      "4 \t 55367316 \t 55516239 \t misteli:target_225\n",
      "4 \t 55439495 \t 55586511 \t misteli:target_225_5\n",
      "4 \t 55586513 \t 55775729 \t misteli:target_226\n",
      "misteli:target_226_5 no start\n",
      "4 \t 56020676 \t 56178582 \t misteli:target_227\n",
      "4 \t 56298731 \t 56485483 \t misteli:target_228\n",
      "noooooooo misteli:target_229\n",
      "noooooooo misteli:target_230\n",
      "4 \t 56986598 \t 57145686 \t misteli:target_231\n",
      "['misteli:target_89_1']\n",
      "\n",
      "['misteli:target_11_1']\n",
      "\n",
      "['misteli:target_89_18']\n",
      "\n",
      "['misteli:target_11_18']\n",
      "\n",
      "['misteli:target_225.5']\n",
      "\n",
      "['misteli:target_614.5']\n",
      "\n",
      "['4dn-dcic-lab:chromosome']\n",
      "\n",
      "['4dn-dcic-lab:Target1']\n",
      "\n"
     ]
    }
   ],
   "source": [
    "als = []\n",
    "for i in prob_num.split('\\n'):\n",
    "    alias = \"misteli:target_\"+i\n",
    "    if '.' in alias:\n",
    "        alias = alias.replace('.', '_')\n",
    "    hit = [i for i in gen if i['aliases'][0] == alias]\n",
    "    if hit:\n",
    "        als.append(alias)\n",
    "        reg = hit[0]['targeted_genome_regions'][0]\n",
    "        try:\n",
    "            print (reg['chromosome'],'\\t',reg['start_coordinate'],'\\t',reg['end_coordinate'],'\\t', alias)\n",
    "        except:\n",
    "            print(alias, 'no start')\n",
    "    else:\n",
    "        pass\n",
    "        print('noooooooo', alias)\n",
    "\n",
    "rem = [i for i in gen if i['aliases'][0] not in als]\n",
    "for i in rem:\n",
    "    print(i['aliases'])\n",
    "    print()\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "4DNFIS1ZVUWO\n",
      "restriction sites for MboI on the fruitfly genome build dm6, juicer-formatted\n",
      "4DNFIO5MGY32\n",
      "This tar-gzipped file includes the bwa index for the fruitfly genome build Release 6 (dm6)\n",
      "4DNFIBEEN92C\n",
      "Chromosome sizes file for the fruitfly genome build Release 6 (dm6); Only the assembled chromosomes are included; unlocalized or unplaced contigs are excluded. \n",
      "4DNFI9PGZ7S6\n",
      "The reference fasta file for the fruitfly genome build Release 6 (dm6)\n",
      "4DNFI0NK4G14\n",
      "restriction sites for MboI on hg38, juicer-formatted\n",
      "4DNFI3HVC1SE\n",
      "restriction sites for DpnII on mm10, juicer-formatted\n",
      "4DNFI3HVU20D\n",
      "restriction sites for NcoII on hg38, juicer-formatted\n",
      "4DNFI3UBJ3HZ\n",
      "Chromosome sizes for mouse genome build mm10; Only the assembled chromosomes are included; unlocalized or unplaced contigs are excluded. \n",
      "4DNFIBNAPW30\n",
      "restriction sites for DpnII on hg38, juicer-formatted\n",
      "4DNFI823MBKE\n",
      "restriction sites for HindIII on hg38, juicer-formatted\n",
      "4DNFI823LSII\n",
      "Chromosome sizes file for the human genome build hg38. Only the assembled chromosomes are included; unlocalized or unplaced contigs are excluded. \n",
      "4DNFI823L812\n",
      "restriction sites for MboI on hg38, juicer-formatted\n",
      "4DNFI823L811\n",
      "restriction sites for HindIII on hg38, juicer-formatted (OLD)\n",
      "4DNFI823LSI8\n",
      "This tar-gzipped file includes the bwa index for the mouse genome build GRCm38. We use the UCSC mm10 version of the genome, corresponsing to GRCm38/GCA_000001635.2,  including the 22 assembled chromosomes (1-19, X, Y, M), the 22 unplaced contigs, and the 22 unlocalized contigs, but excluding any alternative sequences. This is the same reference as used by the ENCODE Consortium for data processing: https://www.encodeproject.org/data-standards/reference-sequences/ .\n",
      "4DNFIZQZ39L9\n",
      "This tar-gzipped file includes the bwa index for the human genome build GRCh38. We use the UCSC hg38 version of the genome, corresponsing to GRCh38/GCA_000001405.15 including the 25 assembled chromosomes (1-22, X, Y, M), the 127 unplaced contigs, the 42 unlocalized contigs, but excluding the 261 alternative haplotypes. The assembly also includes the Epstein-Barr virus (EBV) genome. This is the same reference as used by the ENCODE Consortium for data processing: https://www.encodeproject.org/data-standards/reference-sequences/ .\n"
     ]
    }
   ],
   "source": [
    "text = '''4DNFIS1ZVUWO\trestriction sites for MboI on the fruitfly genome build dm6, juicer-formatted\n",
    "4DNFIO5MGY32\tThis tar-gzipped file includes the bwa index for the fruitfly genome build Release 6 (dm6)\n",
    "4DNFIBEEN92C\tChromosome sizes file for the fruitfly genome build Release 6 (dm6); Only the assembled chromosomes are included; unlocalized or unplaced contigs are excluded. \n",
    "4DNFI9PGZ7S6\tThe reference fasta file for the fruitfly genome build Release 6 (dm6)\n",
    "4DNFI0NK4G14\trestriction sites for MboI on hg38, juicer-formatted\n",
    "4DNFI3HVC1SE\trestriction sites for DpnII on mm10, juicer-formatted\n",
    "4DNFI3HVU20D\trestriction sites for NcoII on hg38, juicer-formatted\n",
    "4DNFI3UBJ3HZ\tChromosome sizes for mouse genome build mm10; Only the assembled chromosomes are included; unlocalized or unplaced contigs are excluded. \n",
    "4DNFIBNAPW30\trestriction sites for DpnII on hg38, juicer-formatted\n",
    "4DNFI823MBKE\trestriction sites for HindIII on hg38, juicer-formatted\n",
    "4DNFI823LSII\tChromosome sizes file for the human genome build hg38. Only the assembled chromosomes are included; unlocalized or unplaced contigs are excluded. \n",
    "4DNFI823L812\trestriction sites for MboI on hg38, juicer-formatted\n",
    "4DNFI823L811\trestriction sites for HindIII on hg38, juicer-formatted (OLD)\n",
    "4DNFI823LSI8\tThis tar-gzipped file includes the bwa index for the mouse genome build GRCm38. We use the UCSC mm10 version of the genome, corresponsing to GRCm38/GCA_000001635.2,  including the 22 assembled chromosomes (1-19, X, Y, M), the 22 unplaced contigs, and the 22 unlocalized contigs, but excluding any alternative sequences. This is the same reference as used by the ENCODE Consortium for data processing: https://www.encodeproject.org/data-standards/reference-sequences/ .\n",
    "4DNFIZQZ39L9\tThis tar-gzipped file includes the bwa index for the human genome build GRCh38. We use the UCSC hg38 version of the genome, corresponsing to GRCh38/GCA_000001405.15 including the 25 assembled chromosomes (1-22, X, Y, M), the 127 unplaced contigs, the 42 unlocalized contigs, but excluding the 261 alternative haplotypes. The assembly also includes the Epstein-Barr virus (EBV) genome. This is the same reference as used by the ENCODE Consortium for data processing: https://www.encodeproject.org/data-standards/reference-sequences/ .'''\n",
    "\n",
    "for t in text.split('\\n'):\n",
    "    i,d = t.split('\t')\n",
    "    ff_utils.patch_metadata({d})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
