{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "from dcicutils import ff_utils\n",
    "from functions.notebook_functions import *\n",
    "from tqdm import tqdm_notebook as tqdm\n",
    "from functions.cleanup import *\n",
    "\n",
    "my_env = 'data'\n",
    "my_key = get_key('koray_data')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "genome = 'None'\n",
    "settings = [{\n",
    "            \"wf_name\": \"encode-chipseq-aln-chip\",\n",
    "            \"wf_uuid\": \"4dn-dcic-lab:wf-encode-chipseq-aln-chip\",\n",
    "            \"parameters\": {},\n",
    "            \"config\": {\n",
    "                       \"ebs_size\": 0,\n",
    "                       \"ebs_type\": \"gp2\",\n",
    "                       \"json_bucket\": \"4dn-aws-pipeline-run-json\",\n",
    "                       \"EBS_optimized\": \"\",\n",
    "                       \"ebs_iops\": \"\",\n",
    "                       \"shutdown_min\": \"now\",\n",
    "                       \"instance_type\": \"\",\n",
    "                       \"password\": \"\",\n",
    "                       \"log_bucket\": \"tibanna-output\",\n",
    "                       \"key_name\": \"\",\n",
    "                       \"cloudwatch_dashboard\": True\n",
    "            },\n",
    "            'custom_pf_fields': {\n",
    "                'chip.first_ta': {\n",
    "                    'genome_assembly': genome,\n",
    "                    'file_type': 'read positions',\n",
    "                    'description': 'Positions of aligned reads in bed format, one line per read mate, for control experiment, from ENCODE ChIP-Seq Pipeline'},\n",
    "                'chip.first_ta_xcor': {\n",
    "                    'genome_assembly': genome,\n",
    "                    'file_type': 'intermediate file',\n",
    "                    'description': 'Counts file used only for QC'}\n",
    "            }\n",
    "        },\n",
    "        {\n",
    "            \"wf_name\": \"encode-chipseq-aln-ctl\",\n",
    "            \"wf_uuid\": \"4dn-dcic-lab:wf-encode-chipseq-aln-ctl\",\n",
    "            \"parameters\": {},\n",
    "            \"config\": {\n",
    "                \"ebs_size\": 0,\n",
    "                \"ebs_type\": \"gp2\",\n",
    "                \"json_bucket\": \"4dn-aws-pipeline-run-json\",\n",
    "                \"EBS_optimized\": \"\",\n",
    "                \"ebs_iops\": \"\",\n",
    "                \"shutdown_min\": 'now',\n",
    "                \"instance_type\": \"\",\n",
    "                \"password\": \"\",\n",
    "                \"log_bucket\": \"tibanna-output\",\n",
    "                \"key_name\": \"\",\n",
    "                \"cloudwatch_dashboard\": True\n",
    "            },\n",
    "            'custom_pf_fields': {\n",
    "                'chip.first_ta_ctl': {\n",
    "                    'genome_assembly': genome,\n",
    "                    'file_type': 'read positions',\n",
    "                    'description': 'Positions of aligned reads in bed format, one line per read mate, for control experiment, from ENCODE ChIP-Seq Pipeline'}\n",
    "            }\n",
    "        },\n",
    "        {\n",
    "            \"wf_name\": \"encode-chipseq-postaln\",\n",
    "            \"wf_uuid\": \"4dn-dcic-lab:wf-encode-chipseq-postaln\",\n",
    "            \"parameters\": {},\n",
    "            \"config\": {\n",
    "                \"ebs_size\": 0,\n",
    "                \"ebs_type\": \"gp2\",\n",
    "                \"json_bucket\": \"4dn-aws-pipeline-run-json\",\n",
    "                \"EBS_optimized\": \"\",\n",
    "                \"ebs_iops\": \"\",\n",
    "                \"shutdown_min\": \"now\",\n",
    "                \"instance_type\": \"\",\n",
    "                \"password\": \"\",\n",
    "                \"log_bucket\": \"tibanna-output\",\n",
    "                \"key_name\": \"\",\n",
    "                \"cloudwatch_dashboard\": True\n",
    "            },\n",
    "            'custom_pf_fields': {\n",
    "                'chip.optimal_peak': {\n",
    "                    'genome_assembly': genome,\n",
    "                    'file_type': 'peaks',\n",
    "                    'description': 'Peak calls from ENCODE ChIP-Seq Pipeline'},\n",
    "                'chip.conservative_peak': {\n",
    "                    'genome_assembly': genome,\n",
    "                    'file_type': 'conservative peaks',\n",
    "                    'description': 'Conservative peak calls from ENCODE ChIP-Seq Pipeline'},\n",
    "                'chip.sig_fc': {\n",
    "                    'genome_assembly': genome,\n",
    "                    'file_type': 'signal fold change',\n",
    "                    'description': 'ChIP-seq signal fold change over input control'}\n",
    "            }\n",
    "        },\n",
    "        {\n",
    "            \"wf_name\": \"encode-atacseq-aln\",\n",
    "            \"wf_uuid\": \"4dn-dcic-lab:wf-encode-atacseq-aln\",\n",
    "            \"parameters\": {},\n",
    "            \"config\": {\n",
    "                \"ebs_size\": 0,\n",
    "                \"ebs_type\": \"gp2\",\n",
    "                \"json_bucket\": \"4dn-aws-pipeline-run-json\",\n",
    "                \"EBS_optimized\": \"\",\n",
    "                \"ebs_iops\": \"\",\n",
    "                \"shutdown_min\": 'now',\n",
    "                \"instance_type\": \"\",\n",
    "                \"password\": \"\",\n",
    "                \"log_bucket\": \"tibanna-output\",\n",
    "                \"key_name\": \"\",\n",
    "                \"cloudwatch_dashboard\": True\n",
    "            },\n",
    "            'custom_pf_fields': {\n",
    "                'atac.first_ta': {\n",
    "                    'genome_assembly': genome,\n",
    "                    'file_type': 'read positions',\n",
    "                    'description': 'Positions of aligned reads in bed format, one line per read mate, from ENCODE ATAC-Seq Pipeline'}\n",
    "            }\n",
    "        },\n",
    "        {\n",
    "            \"wf_name\": \"encode-atacseq-postaln\",\n",
    "            \"wf_uuid\": \"4dn-dcic-lab:wf-encode-atacseq-postaln\",\n",
    "            \"parameters\": {},\n",
    "            \"config\": {\n",
    "                \"ebs_size\": 0,\n",
    "                \"ebs_type\": \"gp2\",\n",
    "                \"json_bucket\": \"4dn-aws-pipeline-run-json\",\n",
    "                \"EBS_optimized\": \"\",\n",
    "                \"ebs_iops\": \"\",\n",
    "                \"shutdown_min\": \"now\",\n",
    "                \"instance_type\": \"\",\n",
    "                \"password\": \"\",\n",
    "                \"log_bucket\": \"tibanna-output\",\n",
    "                \"key_name\": \"\",\n",
    "                \"cloudwatch_dashboard\": True\n",
    "            },\n",
    "            'custom_pf_fields': {\n",
    "                'atac.optimal_peak': {\n",
    "                    'genome_assembly': genome,\n",
    "                    'file_type': 'peaks',\n",
    "                    'description': 'Peak calls from ENCODE ATAC-Seq Pipeline'},\n",
    "                'atac.conservative_peak': {\n",
    "                    'genome_assembly': genome,\n",
    "                    'file_type': 'conservative peaks',\n",
    "                    'description': 'Conservative peak calls from ENCODE ATAC-Seq Pipeline'},\n",
    "                'atac.sig_fc': {\n",
    "                    'genome_assembly': genome,\n",
    "                    'file_type': 'signal fold change',\n",
    "                    'description': 'ATAC-seq signal fold change'}\n",
    "            }\n",
    "        },\n",
    "        {\n",
    "            \"wf_name\": \"mergebed\",\n",
    "            \"wf_uuid\": \"2b10e472-065e-43ed-992c-fccad6417b65\",\n",
    "            \"parameters\": {\"sortv\": \"0\"},\n",
    "            \"config\": {\n",
    "                \"ebs_size\": 0,\n",
    "                \"ebs_type\": \"gp2\",\n",
    "                \"json_bucket\": \"4dn-aws-pipeline-run-json\",\n",
    "                \"EBS_optimized\": \"\",\n",
    "                \"ebs_iops\": \"\",\n",
    "                \"shutdown_min\": \"now\",\n",
    "                \"instance_type\": \"\",\n",
    "                \"password\": \"\",\n",
    "                \"log_bucket\": \"tibanna-output\",\n",
    "                \"key_name\": \"\",\n",
    "                \"cloudwatch_dashboard\": True\n",
    "            }\n",
    "        }\n",
    "    ]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "encode-chipseq-aln-chip\n",
      "+++++++++++++++++++++++++++++++++++++++++++++++++\n",
      "\n",
      "encode-chipseq-aln-ctl\n",
      "+++++++++++++++++++++++++++++\n",
      "\n",
      "encode-chipseq-postaln\n",
      "++++++++++++++++++++++++++\n",
      "\n",
      "encode-atacseq-aln\n",
      "+++++++++++++++++++++++++\n",
      "\n",
      "encode-atacseq-postaln\n",
      "++++++++++++++++\n",
      "\n",
      "mergebed\n",
      "++\n"
     ]
    }
   ],
   "source": [
    "for sett in settings:\n",
    "    print()\n",
    "    print(sett['wf_name'])\n",
    "    url = '/search/?type=WorkflowRunAwsem&workflow.app_name=' + sett['wf_name']\n",
    "    items = ff_utils.search_metadata(url, my_key)\n",
    "    files_to_check = sett['custom_pf_fields']\n",
    "    for i in items:\n",
    "        print('+', end = '')\n",
    "        for f in files_to_check:\n",
    "            my_f = [o['value'] for o in i['output_files'] if o['workflow_argument_name'] == f][0]\n",
    "            patch_data = {'file_type':files_to_check[f]['file_type'].strip(),\n",
    "                          'description':files_to_check[f]['description'].strip()\n",
    "                         }\n",
    "            ff_utils.patch_metadata(patch_data, my_f['uuid'], my_key)\n",
    "    print()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "anaconda-cloud": {},
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
