{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2 exp sets for status change\n",
      "711 items collected\n",
      "5.9 sec for collection\n"
     ]
    }
   ],
   "source": [
    "from dcicutils import ff_utils\n",
    "from functions.notebook_functions import *\n",
    "from functions.cleanup import *\n",
    "import time\n",
    "time1 = time.time()\n",
    "\n",
    "my_auth = get_key('koray_data')\n",
    "\n",
    "# Please Modify the following accordingly \n",
    "# do you want to check for duplicate/problematic runs on files?\n",
    "# it will take some time\n",
    "check_wfrs = False\n",
    "delete_problematic = False\n",
    "\n",
    "# Which status to change\n",
    "change_status = 'released to project'\n",
    "\n",
    "sets_in_scope =  ['4DNESYTWHUH6', '4DNES21D8SP8']\n",
    "search_url = \"\"  # \"/search/type=AAAAAAAA\"\n",
    "\n",
    "\n",
    "if sets_in_scope:\n",
    "    set_to_release = [ff_utils.get_metadata(i, my_auth)['uuid'] for i in sets_in_scope]\n",
    "elif search_url:\n",
    "    set_to_release = [i['uuid'] for i in ff_utils.search_metadata(search_url, my_auth)]\n",
    "\n",
    "store={}\n",
    "item_uuids=[]\n",
    "store, uuids = ff_utils.expand_es_metadata(set_to_release, my_auth, store_frame='object',add_pc_wfr=True, ignore_field = ['experiment_relation', 'biosample_relation', 'references'])\n",
    "\n",
    "print(len(store['experiment_set_replicate']), 'exp sets for status change')\n",
    "print(len(uuids), 'items collected')\n",
    "time2 = time.time()\n",
    "print(round((time2-time1), 1), 'sec for collection')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "EXPSET CHECK\n",
      "\n",
      "EXP CHECK\n",
      "\n",
      "FILE FASTQ CHECK\n",
      "\n",
      "FILE PROCESSED CHECK\n",
      "\n",
      "WFR CHECK\n",
      "  Following run types are found:\n",
      "    pairsqc-single\n",
      "    bwa-mem\n",
      "    hi-c-processing-bam\n",
      "    hi-c-processing-pairs-nore\n",
      "    fastqc-0-11-4-1\n",
      "    md5\n",
      "\n",
      "REPORT NUMBERS AND CHECK STATUS\n",
      "lab 7\n",
      "organism 1\n",
      "ontology 2\n",
      "biosource 2\n",
      "quality_metric_fastqc 92\n",
      "vendor 3\n",
      "ontology_term 15\n",
      "experiment_hi_c 20\n",
      "biosample_cell_culture 4\n",
      "workflow 9\n",
      "individual_human 2\n",
      "file_reference 2\n",
      "image 4\n",
      "software 10\n",
      "static_section 3\n",
      "biosample 4\n",
      "enzyme 1\n",
      "protocol 10\n",
      "file_processed 94\n",
      "file_fastq 92\n",
      "user 15\n",
      "award 6\n",
      "file_format 14\n",
      "quality_metric_pairsqc 23\n",
      "experiment_set_replicate 2\n",
      "workflow_run_awsem 274\n"
     ]
    }
   ],
   "source": [
    "# TODO\n",
    "# Check audits\n",
    "\n",
    "# to decide which items to ignore as they have a 'higher' status\n",
    "STATUS_LEVEL = {\n",
    "    # standard_status\n",
    "    \"released\": 10, \"current\": 10,\n",
    "    \"released to project\": 9,\n",
    "    \"pre-release\": 8,\n",
    "    \"planned\": 6, \"submission in progress\": 6,\n",
    "    \"in review by lab\": 4,\n",
    "    \"revoked\": 0, \"archived\": 0,\"deleted\": 0, \"obsolete\": 0, \"replaced\": 0, \"archived to project\": 0,\n",
    "    # additional file statuses\n",
    "    'to be uploaded by workflow': 4, 'uploading': 4, 'uploaded': 4, 'upload failed': 4, 'draft': 4}\n",
    "\n",
    "change_level = STATUS_LEVEL.get(change_status, 1)\n",
    "\n",
    "# check expsets\n",
    "print('EXPSET CHECK')\n",
    "for a_set in store['experiment_set_replicate']:\n",
    "    if not a_set.get('completed_processes'):\n",
    "        print(a_set['accession'], 'missing processing tag', a_set['description'][:50])\n",
    "\n",
    "# check exps \n",
    "print('\\nEXP CHECK')\n",
    "# check for experiment numbers\n",
    "exp_names = [i for i in store if i.startswith('experiment') and not i.startswith('experiment_set')]\n",
    "all_exps_on_sets = [a for i in store['experiment_set_replicate'] for a in i['experiments_in_set']]\n",
    "all_exps = [a['uuid'] for i in store.keys() for a in store[i] if i in exp_names]\n",
    "if len(all_exps_on_sets) != len(all_exps):\n",
    "    print('Number of experiments is not same as experiments associated with sets')\n",
    "    print('# of exps: {}. # of exps on sets: {}'.format(len(all_exps), len(all_exps_on_sets)))\n",
    "        \n",
    "print('\\nFILE FASTQ CHECK')\n",
    "for a_file in store['file_fastq']:\n",
    "    if not a_file.get('quality_metric'):\n",
    "        print(a_file['accession'], 'missing fastqc')\n",
    "    if not a_file.get('content_md5sum') or not a_file.get('md5sum'):\n",
    "        print(a_file['accession'], 'problems with md5')\n",
    "    if check_wfrs:\n",
    "        try:\n",
    "            # to do skip sbg runs\n",
    "            dw = delete_wfrs(a_file, my_auth, delete=delete_problematic)\n",
    "        except:\n",
    "            print('sbg run skipping')\n",
    "\n",
    "# check processed files\n",
    "print('\\nFILE PROCESSED CHECK')\n",
    "if store.get('file_processed'):\n",
    "    for a_file in store['file_processed']:\n",
    "        if a_file['file_format'] == '/file-formats/pairs/':\n",
    "            if not a_file.get('quality_metric'):\n",
    "                print(a_file['accession'], 'missing Pairsqc')\n",
    "        if not a_file.get('source_experiments'):\n",
    "            print(a_file['accession'], 'missing source experiments')\n",
    "        if check_wfrs:\n",
    "            dw = delete_wfrs(a_file, my_auth, delete=delete_problematic)   \n",
    "\n",
    "# check wfrs\n",
    "print('\\nWFR CHECK')\n",
    "print('  Following run types are found:')\n",
    "for wf in set([i['display_title'].split(' ')[0] for i in store.get('workflow_run_awsem')]):\n",
    "           print('    ' + wf)\n",
    "if store.get('workflow_run_awsem'):\n",
    "    for wfr in store['workflow_run_awsem']:\n",
    "        if wfr['run_status'] != 'complete':\n",
    "            print('problematic wfr', wfr['uuid'], wfr['run_status'])\n",
    "        \n",
    "# check for weird status\n",
    "print('\\nREPORT NUMBERS AND CHECK STATUS')\n",
    "for i in store:\n",
    "    print(i, len(store[i]))\n",
    "    weird = [[i, x['uuid'], x['status']] for x in store[i] if STATUS_LEVEL.get(x['status']) == 0]\n",
    "    if weird:\n",
    "        for case in weird:\n",
    "            print(case)\n",
    "        print()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "lab abd48785-b0e5-4453-be14-30d37a516bf3 Yijun Ruan, JAX\n",
      "user 68b09572-0c3e-48c7-90a9-7bad21d6bac1 Rafael Casellas\n",
      "user d8ac229e-ec08-4411-be38-dc779520ea62 Yijun Ruan\n",
      "award 029a2578-43dc-4343-8f41-694518cce304 NUCLEOME POSITIONING SYSTEM FOR SPATIOTEMPORAL GENOME ORGANIZATION AND REGULATION\n"
     ]
    }
   ],
   "source": [
    "for i in store:\n",
    "    for x in store[i]:\n",
    "        search = str(x)\n",
    "        if 'ruan' in search.lower():\n",
    "            print(i, x['uuid'], x.get('display_title'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "lab                       Out of 7, 7 skipped, 0 matching, 0 unusual, and 0 needs change\n",
      "organism                  Out of 1, 1 skipped, 0 matching, 0 unusual, and 0 needs change\n",
      "ontology                  Out of 2, 2 skipped, 0 matching, 0 unusual, and 0 needs change\n",
      "biosource                 Out of 2, 2 skipped, 0 matching, 0 unusual, and 0 needs change\n",
      "quality_metric_fastqc     Out of 92, 0 skipped, 92 matching, 0 unusual, and 0 needs change\n",
      "vendor                    Out of 3, 3 skipped, 0 matching, 0 unusual, and 0 needs change\n",
      "ontology_term             Out of 15, 15 skipped, 0 matching, 0 unusual, and 0 needs change\n",
      "experiment_hi_c           Out of 20, 0 skipped, 20 matching, 0 unusual, and 0 needs change\n",
      "biosample_cell_culture    Out of 4, 0 skipped, 4 matching, 0 unusual, and 0 needs change\n",
      "workflow                  Out of 9, 9 skipped, 0 matching, 0 unusual, and 0 needs change\n",
      "individual_human          Out of 2, 2 skipped, 0 matching, 0 unusual, and 0 needs change\n",
      "file_reference            Out of 2, 2 skipped, 0 matching, 0 unusual, and 0 needs change\n",
      "image                     Out of 4, 0 skipped, 4 matching, 0 unusual, and 0 needs change\n",
      "software                  Out of 10, 10 skipped, 0 matching, 0 unusual, and 0 needs change\n",
      "static_section            Out of 3, 1 skipped, 0 matching, 0 unusual, and 2 needs change\n",
      "biosample                 Out of 4, 0 skipped, 4 matching, 0 unusual, and 0 needs change\n",
      "enzyme                    Out of 1, 1 skipped, 0 matching, 0 unusual, and 0 needs change\n",
      "protocol                  Out of 10, 2 skipped, 8 matching, 0 unusual, and 0 needs change\n",
      "file_processed            Out of 94, 0 skipped, 0 matching, 0 unusual, and 94 needs change\n",
      "file_fastq                Out of 92, 0 skipped, 92 matching, 0 unusual, and 0 needs change\n",
      "user                      Out of 15, 15 skipped, 0 matching, 0 unusual, and 0 needs change\n",
      "award                     Out of 6, 6 skipped, 0 matching, 0 unusual, and 0 needs change\n",
      "file_format               Out of 14, 14 skipped, 0 matching, 0 unusual, and 0 needs change\n",
      "quality_metric_pairsqc    Out of 23, 0 skipped, 0 matching, 0 unusual, and 23 needs change\n",
      "experiment_set_replicate  Out of 2, 0 skipped, 2 matching, 0 unusual, and 0 needs change\n",
      "workflow_run_awsem        Out of 274, 0 skipped, 120 matching, 0 unusual, and 154 needs change\n"
     ]
    }
   ],
   "source": [
    "# Check status\n",
    "print_each = False\n",
    "counter = 0\n",
    "for a_type in store:\n",
    "    total = len(store[a_type])\n",
    "    change = 0\n",
    "    matching = 0\n",
    "    unusual = 0\n",
    "    skipping = 0\n",
    "    for raw_data in store[a_type]:\n",
    "        if STATUS_LEVEL.get(raw_data['status']) > change_level:\n",
    "            skipping += 1\n",
    "            msg = ('{} {} ITEM HAS STATUS {} HIGHER THAN {} - SKIPPING'.format(a_type, raw_data['uuid'], raw_data['status'], change_status))\n",
    "        elif STATUS_LEVEL.get(raw_data['status']) == STATUS_LEVEL.get(change_status):\n",
    "            matching += 1\n",
    "            msg = ('MATCHING ACCESS STATUS', a_type, raw_data['uuid'], raw_data['status'])\n",
    "        elif STATUS_LEVEL.get(raw_data['status']) == 0:\n",
    "            unusual += 1\n",
    "            msg = ('SKIP UNUSUAL STATUS  ', a_type, raw_data['uuid'], raw_data['status'])\n",
    "        else:\n",
    "            change += 1\n",
    "            msg = ('       CHANGE        ', a_type, raw_data['uuid'], raw_data['status'])\n",
    "        if print_each:\n",
    "            print(msg)\n",
    "    print('{:<25} Out of {t}, {r} skipped, {m} matching, {u} unusual, and {c} needs change'.format(a_type, t=total, r=skipping, m=matching, u=unusual, c=change))\n",
    "    \n",
    "       "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Status will change to released to project\n",
      "Did another wrangler review this release? (y/n):y\n",
      "lab                       Out of 7, 7 skipped, 0 matching, 0 unusual, and 0 UPDATED with status\n",
      "organism                  Out of 1, 1 skipped, 0 matching, 0 unusual, and 0 UPDATED with status\n",
      "ontology                  Out of 2, 2 skipped, 0 matching, 0 unusual, and 0 UPDATED with status\n",
      "biosource                 Out of 2, 2 skipped, 0 matching, 0 unusual, and 0 UPDATED with status\n",
      "quality_metric_fastqc     Out of 92, 0 skipped, 92 matching, 0 unusual, and 0 UPDATED with status\n",
      "vendor                    Out of 3, 3 skipped, 0 matching, 0 unusual, and 0 UPDATED with status\n",
      "ontology_term             Out of 15, 15 skipped, 0 matching, 0 unusual, and 0 UPDATED with status\n",
      "experiment_hi_c           Out of 20, 0 skipped, 20 matching, 0 unusual, and 0 UPDATED with status\n",
      "biosample_cell_culture    Out of 4, 0 skipped, 4 matching, 0 unusual, and 0 UPDATED with status\n",
      "workflow                  Out of 9, 9 skipped, 0 matching, 0 unusual, and 0 UPDATED with status\n",
      "individual_human          Out of 2, 2 skipped, 0 matching, 0 unusual, and 0 UPDATED with status\n",
      "file_reference            Out of 2, 2 skipped, 0 matching, 0 unusual, and 0 UPDATED with status\n",
      "image                     Out of 4, 0 skipped, 4 matching, 0 unusual, and 0 UPDATED with status\n",
      "software                  Out of 10, 10 skipped, 0 matching, 0 unusual, and 0 UPDATED with status\n",
      "static_section            Out of 3, 1 skipped, 0 matching, 0 unusual, and 2 UPDATED with status\n",
      "biosample                 Out of 4, 0 skipped, 4 matching, 0 unusual, and 0 UPDATED with status\n",
      "enzyme                    Out of 1, 1 skipped, 0 matching, 0 unusual, and 0 UPDATED with status\n",
      "protocol                  Out of 10, 2 skipped, 8 matching, 0 unusual, and 0 UPDATED with status\n",
      "file_processed            Out of 94, 0 skipped, 0 matching, 0 unusual, and 94 UPDATED with status\n",
      "file_fastq                Out of 92, 0 skipped, 92 matching, 0 unusual, and 0 UPDATED with status\n",
      "user                      Out of 15, 15 skipped, 0 matching, 0 unusual, and 0 UPDATED with status\n",
      "award                     Out of 6, 6 skipped, 0 matching, 0 unusual, and 0 UPDATED with status\n",
      "file_format               Out of 14, 14 skipped, 0 matching, 0 unusual, and 0 UPDATED with status\n",
      "quality_metric_pairsqc    Out of 23, 0 skipped, 0 matching, 0 unusual, and 23 UPDATED with status\n",
      "experiment_set_replicate  Out of 2, 0 skipped, 2 matching, 0 unusual, and 0 UPDATED with status\n",
      "workflow_run_awsem        Out of 274, 0 skipped, 120 matching, 0 unusual, and 154 UPDATED with status\n"
     ]
    }
   ],
   "source": [
    "# If you want to patch the status, change action to True\n",
    "action = True\n",
    "\n",
    "reviewed = \"\"\n",
    "reviewed = input('Status will change to {}\\nDid another wrangler review this release? (y/n):'.format(change_status))\n",
    "if reviewed != 'y':\n",
    "    raise KeyError('A key step is missing!')\n",
    "    \n",
    "print_each = False\n",
    "counter = 0\n",
    "for a_type in store:\n",
    "    total = len(store[a_type])\n",
    "    change = 0\n",
    "    matching = 0\n",
    "    unusual = 0\n",
    "    skipping = 0\n",
    "    for raw_data in store[a_type]:\n",
    "        if STATUS_LEVEL.get(raw_data['status']) > change_level:\n",
    "            skipping += 1\n",
    "            msg = ('{} {} ITEM HAS STATUS {} HIGHER THAN {} - SKIPPING'.format(a_type, raw_data['uuid'], raw_data['status'], change_status))\n",
    "        elif STATUS_LEVEL.get(raw_data['status']) == STATUS_LEVEL.get(change_status):\n",
    "            matching += 1\n",
    "            msg = ('MATCHING ACCESS STATUS', a_type, raw_data['uuid'], raw_data['status'])\n",
    "        elif STATUS_LEVEL.get(raw_data['status']) == 0:\n",
    "            unusual += 1\n",
    "            msg = ('SKIP UNUSUAL STATUS  ', a_type, raw_data['uuid'], raw_data['status'])\n",
    "        else:\n",
    "            change += 1\n",
    "            msg = ('       CHANGE        ', a_type, raw_data['uuid'], raw_data['status'])\n",
    "            if action:\n",
    "                patch_data = {'status': change_status}\n",
    "                if change_status == 'released' and a_type in ['publication']:\n",
    "                    patch_data = {'status': 'current'}\n",
    "                   \n",
    "                ff_utils.patch_metadata(patch_data, obj_id=raw_data['uuid'] ,key=my_auth) \n",
    "\n",
    "        if print_each:\n",
    "            print(msg)\n",
    "    print('{:<25} Out of {t}, {r} skipped, {m} matching, {u} unusual, and {c} UPDATED with status'.format(a_type, t=total, r=skipping, m=matching, u=unusual, c=change))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "celltoolbar": "Raw Cell Format",
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
